docs/rfcs/0007-multi-tenant-remote-write-in-flow.md

```diff
@@ -0,0 +1,212 @@
+
+The above works fine for GEx. However, there are also Grafana customers who use on-prem hosted databases. They are not able to do multi-tenancy for two reasons:
+* The aforementioned "gateway" component is not open source
+* The agent and the database could be communicating via a non-secure connection. Hence, even if a Gateway is used, deriving tenancy information from authentication details would not be an option.
+
+## Problem statement
+
+In order to support the needs of such on-prem customers we could build support for multi-tenancy within the Grafana Agent. It could retrieve the tenant ID from one of several different sources:
+* The name of a Kubernetes pod.
+* An environment variable.
+* Directly specified in the Agent config (e.g. always using a certain tenant ID for a specific remote write component).
+
+For this to work, somehow the Agent has to keep track of which metrics belong to which tenant, then attach the tenant information in the "X-Scope-OrgID" HTTP header when remote writing them to the DB.
+
+## Potential solutions
+
+### Solution 1 - Support multi-tenancy only in the Gateway instead of the Agent
+
+The simplest solution would be to not support multi-tenancy in the Agent at all. Instead, the Enterprise Gateway component (or parts of it) could be open sourced, which could allow on-prem users to run it on their premises. The Gateway could also be expanded to assign tenancy information based on something other than security credentials.
+
+#### Pros
+
+* No changes to the Agent required.
+* Clear separation of responsibility - only the Gateway manages tenant information.
+* The system architecture of on-prem users would be the same as the enterprise one. Hence, it would be easier to understand and maintain.
+* Good security. Allowing the Agent to set the tenant could mean that an Agent is configured to send data as the wrong tenant (either due to malicious activity or by mistake). Setting tenant information within the Gateway, a server-side component, would be more restricted and secure.
+
+#### Cons
+
+* It is extra work for customers to introduce and operate an extra component.
+* Development work on the Gateway is required in order to open source it and make it more configurable.
+
+### Solution 2 - A generic "sharding" configuration for the Agent remote write
+
+The [```prometheus.remote_write```](https://grafana.com/docs/agent/latest/flow/reference/components/prometheus.remote_write/) Flow component could be expanded to be able to shard the writes into different remote writes under the hood. Each of these shards would have its own write-ahead log (WAL). Also, each shard would be able to remote write to a different endpoint address and with different HTTP header information.
```

Comment:
Note: Caching of series refs is per WAL, so if tenants have a lot of overlap with series, this can lead to ballooning memory usage.

Reply:
Good point, thank you, I added this to the "cons" list. Although not sure if it's a "con" it's more like just something to keep in mind.